packages/algorand-js/src/interpreter/txn.ts

```diff
@@ -0,0 +1,307 @@
+        case TxnField.ConfigAssetTotal: {
+          return assertUint64Defined(tx.assetTotal);
+        }
+        case TxnField.ConfigAssetDecimals: {
+          return assertUint64Defined(tx.assetDecimals);
+        }
+        case TxnField.ConfigAssetDefaultFrozen: {
+          return assertUint64Defined(tx.assetDefaultFrozen);
+        }
+        case TxnField.ConfigAssetUnitName: {
+          return toBytes(tx.assetUnitName);
+        }
+        case TxnField.ConfigAssetName: {
+          return toBytes(tx.assetName);
+        }
+        case TxnField.ConfigAssetURL: {
+          return toBytes(tx.assetURL);
+        }
+        case TxnField.ConfigAssetMetadataHash: {
+          return new Uint8Array(tx.assetMetadataHash);
+        }
+        case TxnField.ConfigAssetManager: {
+          return assertAddr(tx.assetManager);
+        }
+        case TxnField.ConfigAssetReserve: {
+          return assertAddr(tx.assetReserve);
+        }
+        case TxnField.ConfigAssetFreeze: {
+          return assertAddr(tx.assetFreeze);
+        }
+        case TxnField.ConfigAssetClawback: {
+          return assertAddr(tx.assetClawback);
+        }
+        default: {
+          return defaultSpec(txField);
+        }
+      }
+    }
+
+    // Asset Transfer Transaction
+    case 'axfer': {
+      switch (txField) {
+        case TxnField.XferAsset: {
+          return assertUint64Defined(tx.assetIndex);
+        }
+        case TxnField.AssetAmount: {
+          return assertUint64Defined(tx.amount);
+        }
+        case TxnField.AssetSender: {
+          return assertAddr(tx.from);
+        }
+        case TxnField.AssetReceiver: {
+          return assertAddr(tx.to);
+        }
+        case TxnField.AssetCloseTo: {
+          return assertAddr(tx.closeRemainderTo);
+        }
+        default: {
+          return defaultSpec(txField);
+        }
+      }
+    }
+
+    // Asset Freeze Transaction
+    case 'afrz': {
+      switch (txField) {
+        case TxnField.FreezeAsset: {
+          return assertUint64Defined(tx.assetIndex);
+        }
+        case TxnField.FreezeAssetAccount: {
+          return assertAddr(tx.freezeAccount);
+        }
+        case TxnField.FreezeAssetFrozen: {
+          return assertUint64Defined(tx.freezeState);
+        }
+        default: {
+          return defaultSpec(txField);
+        }
+      }
+    }
+
+    // Application Call Transaction
+    case 'appl': {
+      switch (txField) {
+        case TxnField.ApplicationID: {
+          return assertUint64Defined(tx.appIndex);
+        }
+        case TxnField.OnCompletion: {
+          return assertUint64Defined(tx.appOnComplete);
+        }
+        case TxnField.ApprovalProgram: {
+          return assertBytesDefined(tx.appApprovalProgram);
+        }
+        case TxnField.ClearStateProgram: {
+          return assertBytesDefined(tx.appClearProgram);
+        }
+        default: {
+          return defaultSpec(txField);
+        }
+      }
+    }
```

Comment:
I would prefer to use the `get_obj_for_encoding` as noted in the Discord chat.
This way we don't need to do that big switch, and if the `Transaction` object change (which is internal to the SDK, NOT the Algorand spec) our code will still work.
So the flow would be:

```
// in interpreter (constructor, or other method where we set the transactions):
this.tx = txSDK.get_obj_for_encoding()
```
TxnField can be defined as a map instead of enum in `types.ts`:
```
const TxnFields = {
   Amount: "amt"
   AssetAmount: "aamt"
   ...
}
// now we can declare a valid type for the TxField:
type TxField = keyof TxFields

const TxFieldDefaults: {[key: TxField]: any}  = {
  Amount: zeroUint64
  ...
}
```
Execute is much simpler and we don't need `txnSpecbyField`:
```
export class Txna extends Op {
   ...
   field: string;

  constructor(field: TxField, .... ) {
    ...
    this.field = types.TxnFields[field] ;
  }

  execute (stack: TEALStack): void {
    const r = this.interpreter.tx[this.field]  || types.TxFieldDefaults[this.field];
    stack.push(r);
  }
}

Reply:
Let me know what do you think. We can keep your implementation - it also works ;)

Reply:
1. Agree about using map instead of enums, `get_obj_for_encoding()`.
2.  But i still think we would need `txnSpecbyField()` function as there are some TxFields which we'll need to handle (eg `TypeEnum`, `GroupIndex`, `TxID`, `NumAppArgs`..)